File: mantra_dev/mantra_dev/report/delivery_note_report/delivery_note_report.js
frappe.query_reports["Delivery Note Report"] = {
    "filters": [
        {
            "fieldname": "customer",
            "label": "Customer",
            "fieldtype": "Link",
            "options": "Customer",
            "width": 300
        },
        {
            "fieldname": "custom_sales_invoice_no",
            "label": __("Sales Invoice"),
            "fieldtype": "Link",
            "options": "Sales Invoice",
            "width": 300
        }
    ],
	onload: function(report) {
        // frappe.query_report.set_filter_value('from_date', frappe.datetime.month_start());
        // frappe.query_report.set_filter_value('to_date', frappe.datetime.month_end());
        
        $(document).on('click', '.create-shipment', function() {
            var data = $(this).data();
            var delivery_note_no = data.id;
			
			console.log(data); // Log the data attributes for debugging
			let d = new frappe.ui.Dialog({
                title: 'Shipment Form',
                fields: [
                    {
                        fieldtype: 'Section Break',
                        label: 'Shipment Details'
                    },
                    {
                        label: 'Delivery Note ID',
                        fieldname: 'dc_id',
                        fieldtype: 'Link',
                        options: "Delivery Note",
                        reqd: 1,
						default:delivery_note_no,
						read_only:1,
                        get_query: function() {
                            return {
                                filters: [
                                    ['Delivery Note', 'docstatus', '=', 1]
                                ]
                            };
                        }
                    },
                    {
                        label: 'Service Provider',
                        fieldname: 'service_provider',
                        fieldtype: 'Data',
                        reqd: 1,
                    },
                    {
                        fieldtype: 'Column Break'
                    },
                    {
                        label: 'AWB Number',
                        fieldname: 'awb_number',
                        fieldtype: 'Data',
                        reqd: 1,
                    },
                    
                ],
                size: 'small', // small, large, extra-large 
                primary_action_label: 'Submit',
                primary_action(values) {
                    console.log(values);
                    // Add code to handle the submission
                    frappe.call({
                        method: "mantra_dev.backend_code.api.create_shipment",
                        args: {
                            values: values,
                            delivery_note_id: values.dc_id,
                            service_provider: values.service_provider,
                            awb_number: values.awb_number,
                            attachment: values.attachment,
                            pickup_from: values.pickup_from,
                            pickup_to: values.pickup_to,
                            shipment_parcel: values.shipment_parcel
                        },
                        callback: function(response) {
                            if (response.message) {
                                frappe.msgprint(__('Shipment created successfully!'));
                            } else {
                                frappe.msgprint(__('An error occurred while creating the shipment.'));
                            }
                        }
                    });
                    d.hide();
                }
            });
            d.show();
        });
	},
	formatter: function(value, row, column, data, default_formatter) {
			if (column.fieldname == "shipment") {
				return value;
			}
			return default_formatter(value, row, column, data);
	}
}

File: mantra_dev/mantra_dev/report/delivery_note_report/delivery_note_report.py
from __future__ import unicode_literals
import frappe
from frappe import _

def execute(filters=None):
    columns, data = get_columns(), get_data(filters)
    return columns, data

def get_columns():
    return [
        {"label": _("Shipment"), "fieldname": "shipment", "fieldtype": "Data"},	
        {"label": _("Delivery Note No"), "fieldname": "delivery_note_no", "fieldtype": "Link", "options": "Delivery Note", "width": 200},
        {"label": _("Custom Sales Invoice No"), "fieldname": "custom_sales_invoice_no", "fieldtype": "Data", "width": 200},
        {"label": _("Customer Name"), "fieldname": "customer_name", "fieldtype": "Data", "width": 200},
        {"label": _("Posting Date"), "fieldname": "posting_date", "fieldtype": "Date", "width": 120},
        {"label": _("Document Status"), "fieldname": "docstatus", "fieldtype": "Int", "width": 100},
        {"label": _("Shipping Address Name"), "fieldname": "shipping_address_name", "fieldtype": "Data", "width": 200},
        {"label": _("Shipping Address"), "fieldname": "shipping_address", "fieldtype": "Data", "width": 300},
        {"label": _("Dispatch Address Name"), "fieldname": "dispatch_address_name", "fieldtype": "Data", "width": 200},
        {"label": _("Dispatch Address"), "fieldname": "dispatch_address", "fieldtype": "Data", "width": 300},
        
    ]

def get_data(filters):
    conditions = ""
    if filters.get("customer"):
        conditions += " AND customer_name = %(customer)s"
    if filters.get("custom_sales_invoice_no"):
        conditions += " AND custom_sales_invoice_no = %(custom_sales_invoice_no)s"

    data = frappe.db.sql("""
        SELECT
            name AS delivery_note_no,
            custom_sales_invoice_no AS custom_sales_invoice_no,
            customer_name AS customer_name,
            posting_date AS posting_date,
            docstatus AS docstatus,
            shipping_address_name AS shipping_address_name,
            shipping_address AS shipping_address,
            dispatch_address_name AS dispatch_address_name,
            dispatch_address AS dispatch_address
        FROM
            `tabDelivery Note`
        WHERE
            docstatus = 1 and custom_shipment_createed=0
            {conditions}
        ORDER BY
            posting_date DESC
    """.format(conditions=conditions), filters, as_dict=True)
    
    for row in data:
        row["shipment"] = '<button class="btn btn-primary pt-0 pb-0 create-shipment" data-id="{}">Shipment</button>'.format(row["delivery_note_no"])
    
    return data
@frappe.whitelist()
def create_shipment(delivery_note_no):
    dc_id=frappe.get_doc("Delivery Note",delivery_note_no)
    dc_id
    frappe.msgprint("function called")
    return "Done"